File: artifacts/ht/translate/src/dump.rs
use translate_new::*;
use std::env;
use std::fmt;
pub fn main() {
    let items = vec![
        Item {
            key: Some("foo".to_string()),
            value: 10,
        },
        Item {
            key: Some("bar".to_string()),
            value: 42,
        },
        Item {
            key: Some("bazz".to_string()),
            value: 36,
        },
        Item {
            key: Some("buzz".to_string()),
            value: 7,
        },
        Item {
            key: Some("bob".to_string()),
            value: 11,
        },
        Item {
            key: Some("jane".to_string()),
            value: 100,
        },
        Item {
            key: Some("x".to_string()),
            value: 200,
        },
    ];
    let num_items = items.len();

    let mut table = match ht_create() {
        Some(t) => t,
        None => {
            exit_nomem();
            return;
        }
    };

    for i in 0..num_items {
        let item = &items[i];
        if ht_set(
            &mut table,
            item.key.as_deref(),
            Some(Box::new(item.value) as Box<dyn std::any::Any>),
        ).is_none()
        {
            exit_nomem();
            return;
        }
    }

    if let Some(entries) = &table.entries {
        for i in 0..table.capacity {
            if i < entries.len() {
                let entry = &entries[i];
                if let Some(key) = &entry.key {
                    if let Some(val) = entry.value.as_ref().and_then(|v| v.downcast_ref::<i32>()) {
                        println!("index {}: key {}, value {}", i, key, val);
                    } else {
                        println!("index {}: key {}, value (invalid type)", i, key);
                    }
                } else {
                    println!("index {}: empty", i);
                }
            } else {
                println!("index {}: empty", i);
            }
        }
    } else {
        for i in 0..table.capacity {
            println!("index {}: empty", i);
        }
    }
}

File: artifacts/ht/decompose/src/dump.c
int main(void)
{
  item items[] = {{"foo", 10}, {"bar", 42}, {"bazz", 36}, {"buzz", 7}, {"bob", 11}, {"jane", 100}, {"x", 200}};
  size_t num_items = (sizeof(items)) / (sizeof(item));
  ht *table = ht_create();
  if (table == 0)
  {
    exit_nomem();
  }
  for (int i = 0; i < num_items; i += 1)
  {
    if (ht_set(table, items[i].key, &items[i].value) == 0)
    {
      exit_nomem();
    }
  }

  for (int i = 0; i < table->capacity; i += 1)
  {
    if (table->entries[i].key != 0)
    {
      printf("index %d: key %s, value %d\n", i, table->entries[i].key, *((int *) table->entries[i].value));
    }
    else
    {
      printf("index %d: empty\n", i);
    }
  }

  return 0;
}


/*** DEPENDENCIES:
ht *ht_create(void)
{
  ht *table = malloc(sizeof(ht));
  if (table == 0)
  {
    return 0;
  }
  table->length = 0;
  table->capacity = 16;
  table->entries = calloc(table->capacity, sizeof(ht_entry));
  if (table->entries == 0)
  {
    free(table);
    return 0;
  }
  return table;
}


----------------------------
typedef struct 
{
  char *key;
  int value;
} item
----------------------------
struct ht
{
  ht_entry *entries;
  size_t capacity;
  size_t length;
}
----------------------------
void exit_nomem(void)
{
  fprintf(stderr, "out of memory\n");
  exit(1);
}


----------------------------
const char *ht_set(ht *table, const char *key, void *value)
{
  assert(value != 0);
  if (value == 0)
  {
    return 0;
  }
  if (table->length >= (table->capacity / 2))
  {
    if (!ht_expand(table))
    {
      return 0;
    }
  }
  return ht_set_entry(table->entries, table->capacity, key, value, &table->length);
}


----------------------------
***/


